tests/integration-tests/tests/truncation_tests.rs

```diff
@@ -0,0 +1,164 @@
+use std::collections::BTreeMap;
+use std::net::{Ipv4Addr, SocketAddr, SocketAddrV4};
+use std::str::FromStr;
+use std::sync::Arc;
+use tokio::net::UdpSocket;
+use trust_dns_client::client::AsyncClient;
+use trust_dns_proto::op::{Edns, Message, MessageType, OpCode, Query};
+use trust_dns_proto::rr::rdata::{A, SOA};
+use trust_dns_proto::rr::{DNSClass, Name, RData, Record, RecordSet, RecordType, RrKey};
+use trust_dns_proto::udp::UdpClientStream;
+use trust_dns_proto::xfer::FirstAnswer;
+use trust_dns_proto::DnsHandle;
+use trust_dns_server::authority::{Catalog, ZoneType};
+use trust_dns_server::store::in_memory::InMemoryAuthority;
+use trust_dns_server::ServerFuture;
+
+#[tokio::test]
+async fn test_truncation() {
+    let _guard = subscribe();
+
+    let addr = SocketAddr::V4(SocketAddrV4::new(Ipv4Addr::new(127, 0, 0, 1), 0));
+    let udp_socket = UdpSocket::bind(&addr).await.unwrap();
+
+    let nameserver = udp_socket.local_addr().unwrap();
+    println!("udp_socket on port: {nameserver}");
+
+    // Create and start the server.
+    let mut server = ServerFuture::new(new_large_catalog(128));
+    server.register_socket(udp_socket);
+    tokio::spawn(server.block_until_done());
+
+    // Create the UDP client.
+    let stream = UdpClientStream::<UdpSocket>::new(nameserver);
+    let (mut client, bg) = AsyncClient::connect(stream).await.unwrap();
+
+    // Run the client exchange in the background.
+    tokio::spawn(bg);
+
+    // Build the query.
+    let max_payload = 512;
+    let mut msg = Message::new();
+    msg.add_query({
+        let mut query = Query::query(large_name(), RecordType::A);
+        query.set_query_class(DNSClass::IN);
+        query
+    })
+    .set_id(rand::random::<u16>())
+    .set_message_type(MessageType::Query)
+    .set_op_code(OpCode::Query)
+    .set_recursion_desired(true)
+    .set_edns({
+        let mut edns = Edns::new();
+        edns.set_max_payload(max_payload).set_version(0);
+        edns
+    });
+
+    let result = client.send(msg).first_answer().await.expect("query failed");
+
+    assert!(result.truncated());
+    assert_eq!(max_payload, result.max_payload());
+}
+
+// TODO: should we do this for all of the integration tests?
+fn subscribe() -> tracing::subscriber::DefaultGuard {
```

Comment:
@bluejekyll any thoughts about making this a common utility and using it throughout the trust-dns integration tests?

Reply:
Yes, perhaps. Though not as part of this PR.